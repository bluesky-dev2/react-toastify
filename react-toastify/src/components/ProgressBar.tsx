import React, { CSSProperties } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { TYPE, RT_NAMESPACE, TypeOptions } from './../utils';

export interface ProgressBarProps {
  delay: number;
  isRunning: boolean;
  closeToast: () => void;
  type: TypeOptions;
  hide?: boolean;
  className?: string;
  style?: CSSProperties;
  controlledProgress?: boolean;
  progress?: number;
  rtl?: boolean;
}

export function ProgressBar({
  delay,
  isRunning,
  closeToast,
  type,
  hide,
  className,
  style: userStyle,
  controlledProgress,
  progress,
  rtl
}: ProgressBarProps) {
  const style: CSSProperties = {
    ...userStyle,
    animationDuration: `${delay}ms`,
    animationPlayState: isRunning ? 'running' : 'paused',
    opacity: hide ? 0 : 1
  };

  if (controlledProgress) style.transform = `scaleX(${progress})`;

  const classNames = cx(
    `${RT_NAMESPACE}__progress-bar`,
    controlledProgress
      ? `${RT_NAMESPACE}__progress-bar--controlled`
      : `${RT_NAMESPACE}__progress-bar--animated`,
    `${RT_NAMESPACE}__progress-bar--${type}`,
    {
      [`${RT_NAMESPACE}__progress-bar--rtl`]: rtl
    },
    className
  );

  // 🧐 controlledProgress is derived from progress
  // so if controlledProgress is set
  // it means that this is also the case for progress
  const animationEvent = {
    [controlledProgress && progress! >= 1
      ? 'onTransitionEnd'
      : 'onAnimationEnd']:
      controlledProgress && progress! < 1 ? null : closeToast
  };

  return <div className={classNames} style={style} {...animationEvent} />;
}

ProgressBar.propTypes = {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]).isRequired,

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: PropTypes.bool.isRequired,

  /**
   * Func to close the current toast
   */
  closeToast: PropTypes.func.isRequired,

  /**
   * Support rtl content
   */
  rtl: PropTypes.bool.isRequired,

  /**
   * Optional type : info, success ...
   */
  type: PropTypes.string,

  /**
   * Hide or not the progress bar
   */
  hide: PropTypes.bool,

  /**
   * Optionnal className
   */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * Controlled progress value
   */
  progress: PropTypes.number,

  /**
   * Tell wether or not controlled progress bar is used
   */
  controlledProgress: PropTypes.bool
};

ProgressBar.defaultProps = {
  type: TYPE.DEFAULT,
  hide: false
};
