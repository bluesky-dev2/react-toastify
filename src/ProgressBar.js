import React from 'react';
import PropTypes from 'prop-types';
import { css } from "glamor";

import { TYPE } from './constant';
import style from "./style";

const trackProgress = css.keyframes('track-progress', {
  "0%": { width: "100%" },
  "100%": { width: 0 }
});

const progress = type => css({
  position: "absolute",
  bottom: 0,
  left: 0,
  width: 0,
  height: "5px",
  zIndex: 999,
  opacity: 0.7,
  animation: `${trackProgress} linear 1`,
  backgroundColor: "rgba(255,255,255,.7)",
  ...type === "default" ? { background: style.colorProgressDefault } : {}
});

function ProgressBar({ delay, isRunning, closeToast, type, hide, className }) {
  const style = {
    animationDuration: `${delay}ms`,
    animationPlayState: isRunning ? 'running' : 'paused'
  };
  style.WebkitAnimationPlayState = style.animationPlayState;

  if (hide) {
    style.opacity = 0;
  }

  return (
    <div
      {...progress(type)}
      className={className}
      style={style}
      onAnimationEnd={closeToast}
    />
  );
}

ProgressBar.propTypes = {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: PropTypes.number.isRequired,

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: PropTypes.bool.isRequired,

  /**
   * Func to close the current toast
   */
  closeToast: PropTypes.func.isRequired,

  /**
   * Optional type : info, success ...
   */
  type: PropTypes.string,

  /**
   * Hide or not the progress bar
   */
  hide: PropTypes.bool,

  /**
   * Optionnal className
   */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
};

ProgressBar.defaultProps = {
  type: TYPE.DEFAULT,
  hide: false,
  className: ''
};

export default ProgressBar;
