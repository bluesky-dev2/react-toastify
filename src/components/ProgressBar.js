import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { TYPE } from './../utils/constant';

function ProgressBar({
  delay,
  isRunning,
  closeToast,
  type,
  hide,
  className,
  rtl
}) {
  const style = {
    animationDuration: `${delay}ms`,
    animationPlayState: isRunning ? 'running' : 'paused',
    opacity: hide ? 0 : 1
  };
  style.WebkitAnimationPlayState = style.animationPlayState;

  const classNames = cx(
    'react-toastify__progress-bar',
    `react-toastify__progress-bar--${type}`,
    className,
    {
      'react-toastify__progress-bar--rtl': rtl
    }
  );

  return (
    <div className={classNames} style={style} onAnimationEnd={closeToast} />
  );
}

ProgressBar.propTypes = {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: PropTypes.number.isRequired,

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: PropTypes.bool.isRequired,

  /**
   * Func to close the current toast
   */
  closeToast: PropTypes.func.isRequired,

  /**
   * Support rtl content
   */
  rtl: PropTypes.bool.isRequired,

  /**
   * Optional type : info, success ...
   */
  type: PropTypes.string,

  /**
   * Hide or not the progress bar
   */
  hide: PropTypes.bool,

  /**
   * Optionnal className
   */
  className: PropTypes.string
};

ProgressBar.defaultProps = {
  type: TYPE.DEFAULT,
  hide: false,
};

export default ProgressBar;
