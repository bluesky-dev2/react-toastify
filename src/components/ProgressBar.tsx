import React, { CSSProperties } from 'react';
import cx from 'classnames';

import { TYPE, RT_NAMESPACE, TypeOptions } from './../utils';

export interface ProgressBarProps {
  /**
   * The animation delay which determine when to close the toast
   */
  delay: number;

  /**
   * Whether or not the animation is running or paused
   */
  isRunning: boolean;

  /**
   * Func to close the current toast
   */
  closeToast: () => void;

  /**
   * Optional type : info, success ...
   */
  type: TypeOptions;

  /**
   * Hide or not the progress bar
   */
  hide?: boolean;

  /**
   * Optionnal className
   */
  className?: string | null;

  /**
   * Optionnal inline style
   */
  style?: CSSProperties;

  /**
   * Tell wether or not controlled progress bar is used
   */
  controlledProgress?: boolean;

  /**
   * Controlled progress value
   */
  progress?: number | string;

  /**
   * Support rtl content
   */
  rtl?: boolean;
}

export function ProgressBar({
  delay,
  isRunning,
  closeToast,
  type,
  hide,
  className,
  style: userStyle,
  controlledProgress,
  progress,
  rtl
}: ProgressBarProps) {
  const style: CSSProperties = {
    ...userStyle,
    animationDuration: `${delay}ms`,
    animationPlayState: isRunning ? 'running' : 'paused',
    opacity: hide ? 0 : 1
  };

  if (controlledProgress) style.transform = `scaleX(${progress})`;

  const classNames = cx(
    `${RT_NAMESPACE}__progress-bar`,
    controlledProgress
      ? `${RT_NAMESPACE}__progress-bar--controlled`
      : `${RT_NAMESPACE}__progress-bar--animated`,
    `${RT_NAMESPACE}__progress-bar--${type}`,
    {
      [`${RT_NAMESPACE}__progress-bar--rtl`]: rtl
    },
    className
  );

  // 🧐 controlledProgress is derived from progress
  // so if controlledProgress is set
  // it means that this is also the case for progress
  const animationEvent = {
    [controlledProgress && progress! >= 1
      ? 'onTransitionEnd'
      : 'onAnimationEnd']:
      controlledProgress && progress! < 1 ? null : closeToast
  };

  return <div className={classNames} style={style} {...animationEvent} />;
}

ProgressBar.defaultProps = {
  type: TYPE.DEFAULT,
  hide: false
};
