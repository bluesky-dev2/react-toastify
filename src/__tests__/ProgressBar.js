/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';

import ProgressBar from './../ProgressBar';

const REQUIRED_PROPS = {
  delay: 5000,
  isRunning: true,
  closeToast: jest.fn()
};

describe('ProgressBar', () => {
  it('Should merge className', () => {
    const component = shallow(
      <ProgressBar
        {...REQUIRED_PROPS}
        className="test"
      />
    );
    expect(component.find('.test')).toHaveLength(1);
  });

  it('Should call closeToast function when animation end', () => {
    const component = shallow(<ProgressBar {...REQUIRED_PROPS} />);

    expect(REQUIRED_PROPS.closeToast).not.toHaveBeenCalled();
    component.simulate('animationEnd');
    expect(REQUIRED_PROPS.closeToast).toHaveBeenCalled();
  });
});
