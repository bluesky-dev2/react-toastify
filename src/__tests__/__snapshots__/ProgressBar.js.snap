// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Should allow glamor rule as className 1`] = `
.css-11myqvn,
[data-css-11myqvn] {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 5px;
  z-index: 9999;
  opacity: 0.7;
  animation: animation_91emau linear 1;
  animation-play-state: running;
  animation-duration: 5000ms;
  background-color: rgba(255,255,255,.7);
  background: red;
  -webkit-animation: animation_91emau linear 1;
  -webkit-animation-play-state: running;
  -webkit-animation-duration: 5000ms;
}

<div
  data-css-11myqvn=""
  onAnimationEnd={[Function]}
/>
`;

exports[`ProgressBar Should be able to hide the progress bar 1`] = `
.css-7lq92a,
[data-css-7lq92a] {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 5px;
  z-index: 9999;
  opacity: 0;
  animation: animation_91emau linear 1;
  animation-play-state: running;
  animation-duration: 5000ms;
  background-color: rgba(255,255,255,.7);
  background: -webkit-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: -moz-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  -webkit-animation: animation_91emau linear 1;
  -webkit-animation-play-state: running;
  -webkit-animation-duration: 5000ms;
}

<div
  className=""
  data-css-7lq92a=""
  onAnimationEnd={[Function]}
/>
`;

exports[`ProgressBar Should be able to pause animation 1`] = `
.css-9z8rkz,
[data-css-9z8rkz] {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 5px;
  z-index: 9999;
  opacity: 0.7;
  animation: animation_91emau linear 1;
  animation-play-state: paused;
  animation-duration: 5000ms;
  background-color: rgba(255,255,255,.7);
  background: -webkit-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: -moz-linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  background: linear-gradient(to right, #4cd964, #5ac8fa, #007aff, #34aadc, #5856d6, #ff2d55);
  -webkit-animation: animation_91emau linear 1;
  -webkit-animation-play-state: paused;
  -webkit-animation-duration: 5000ms;
}

<div
  className=""
  data-css-9z8rkz=""
  onAnimationEnd={[Function]}
/>
`;
